adapters/yeahmobi/yeahmobi.go

```diff
@@ -0,0 +1,175 @@
+package yeahmobi
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"net/http"
+	"net/url"
+	"text/template"
+)
+
+type YeahmobiAdapter struct {
+	EndpointTemplate template.Template
+}
+
+func NewYeahmobiBidder(endpointTemplate string) adapters.Bidder {
+	tpl, err := template.New("endpointTemplate").Parse(endpointTemplate)
+	if err != nil {
+		glog.Fatal("Unable parse url template err:" + err.Error())
+		return nil
+	}
+	return &YeahmobiAdapter{EndpointTemplate: *tpl}
+}
+
+func (adapter *YeahmobiAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var adapterRequests []*adapters.RequestData
+
+	adapterRequest, errs := adapter.makeRequest(request)
+	if errs == nil {
+		adapterRequests = append(adapterRequests, adapterRequest)
+	}
+
+	return adapterRequests, errs
+}
+
+func (adapter *YeahmobiAdapter) makeRequest(request *openrtb.BidRequest) (*adapters.RequestData, []error) {
+	var errs []error
+
+	yeahmobiExt, errs := getYeahmobiExt(request)
+
+	if yeahmobiExt == nil {
+		glog.Fatal("Invalid ExtImpYeahmobi value")
+		return nil, errs
+	}
+	endPoint, err := adapter.getEndpoint(yeahmobiExt)
+	if err != nil {
+		return nil, append(errs, err)
+	}
+	transform(request)
+	reqBody, err := json.Marshal(request)
+
+	if err != nil {
+		return nil, append(errs, err)
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     endPoint,
+		Body:    reqBody,
+		Headers: headers,
+	}, errs
+}
+
+func transform(request *openrtb.BidRequest) {
+	for i, imp := range request.Imp {
+		if imp.Native != nil {
+			var nativeRequest map[string]interface{}
+			err := json.Unmarshal([]byte(request.Imp[i].Native.Request), &nativeRequest)
+			if err == nil {
+				_, exists := nativeRequest["native"]
+				if exists {
+					continue
+				}
+				nativeCopy := *request.Imp[i].Native
+				nativeCopy.Request = "{\"native\":" + request.Imp[i].Native.Request + "}"
```

Comment:
Would it be a good idea to run this through the json.Marshal? .. or can you guarantee that wrapping `request.Imp[i].Native.Request` in quotes always produces valid json?

Reply:
Yes.It's a good idea and I optimized it in the new commit.